### PR TITLE
feat(templates): process Feature templates in memory

### PR DESCRIPTION
**Pull request by cam-garrison**
## Description
This PR changes our approach with templates to now process/store the templates in memory, rather than copying .tmpl files to /tmp/.... and then creating the .yaml files there as well.

This means that when creating a `feature.Manifests(path)`, you no longer will include the "rootdir" or similar variable in the path. See PR changes for examples of this, since you point directly to the path in /templates.

In addition, when creating a Feature which points to Manifests, you can specify a separate ManifestSource() which takes in a file system. See the tests for examples of pointing to various ManifestSources. 

## How Has This Been Tested?

I've added additional tests to the features test suite to ensure that one is able to apply both .yaml and .tmpl manifests, as well as manifests from default and non-default manifest sources.

In addition, manually tested with image `quay.io/cgarriso/opendatahub-operator:templates-incubation`, creating a DSCI and ensuring that manifests used to create SMCP were still applied as expected.

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: pkg/feature/manifest.go [hunk 1]
```diff
@@ -1,76 +1,103 @@
...
-			name:     basePath,
-			path:     path,
-			patch:    strings.Contains(basePath, ".patch"),
-			template: filepath.Ext(path) == ".tmpl",
-		})
+
+		_, err = fs.ReadFile(fsys, path)
+		if err != nil {
+			return err
+		}
```
Comment (bartoszmajsak) [inline pkg/feature/manifest.go:40]: I do not think we need to read the file here, as we anyway do not use the content. To optimize it, I was thinking we can do the following instead:

```golang
_, err = dirEntry.Info()
if err != nil {
	return err
}

if dirEntry.IsDir() {
	return nil
}
```

(I can't really `suggest` this as it changes the order of already existing code)

Calling `Info` will result in calling `fstat` underneath. This is enough to capture errors such as "file not found", "permission denied", "general i/o" in advance.

File: pkg/feature/manifest.go [hunk 2]
```diff
@@ -1,76 +1,103 @@
...
+	})
+
+	if err != nil {
+		return nil, err
 	}
 
 	return manifests, nil
 }
 
+func loadManifestFrom(path string) manifest {
```
Comment (bartoszmajsak) [inline pkg/feature/manifest.go:54]: `load` here is a bit misleading as we are not loading anything yet at this point... maybe simply `createManifestFrom` is a bit better.

File: pkg/feature/manifest.go [hunk 3]
```diff
@@ -1,76 +1,103 @@
...
-
+		log.Error(err, "Failed to read manifest file", "path", m.path)
 		return err
 	}
 
-	tmpl := template.New(m.name).Funcs(template.FuncMap{"ReplaceChar": ReplaceChar})
+	if !m.template {
+		m.processedContent = string(content)
+		return nil
+	}
```
Comment (bartoszmajsak) [inline pkg/feature/manifest.go:87]: Perhaps a little bit of "why" can be helpful

```suggestion
	if !m.template {
		// If, by convention, the file is not suffixed with `.tmpl` we do not need to trigger template processing.
		// It's safe to return at this point.
		m.processedContent = string(content)
		return nil
	}
```

File: tests/integration/features/features_int_test.go [hunk 4]
```diff
@@ -127,6 +135,119 @@ var _ = Describe("preconditions", func() {
...
+
+			Expect(err).ToNot(HaveOccurred())
+
+			// when
+			By("verifying service is created after applying feature", func() {
+				Expect(createService.Apply()).To(Succeed())
+				service, err := getService("knative-local-gateway", "service-ns")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(service).ToNot(BeNil())
+			})
```
Comment (bartoszmajsak) [inline tests/integration/features/features_int_test.go:175]: The `By` clause here is a bit confusing. If we structure the tests using "gherkin" BDD blocks, then we miss the verification block:

* `given` - setting the test fixture
* `when` - exercise system under test
* `then` - check if your expectations match the reality

So maybe this instead:

```suggestion
			// when
			Expect(createService.Apply()).To(Succeed())

			// then
			service, err := getService("knative-local-gateway", "service-ns")
			Expect(err).ToNot(HaveOccurred())
			Expect(service).ToNot(BeNil())
```

How about testing the service name instead of just its existence (`Expect(service).ToNot(BeNil())`). This would to some extent ensure that the template was processed.

File: tests/integration/features/features_int_test.go [hunk 5]
```diff
@@ -127,6 +135,119 @@ var _ = Describe("preconditions", func() {
...
+		})
+
+		It("should be able to process an embedded yaml file from the default location", func() {
```
Comment (bartoszmajsak) [inline tests/integration/features/features_int_test.go:178]: (nit)

```suggestion
		It("should be able to process an embedded YAML file from the default location", func() {
```

File: tests/integration/features/features_int_test.go [hunk 6]
```diff
@@ -127,6 +135,119 @@ var _ = Describe("preconditions", func() {
...
+
+			Expect(err).ToNot(HaveOccurred())
+
+			// when
+			By("verifying namespace is created after applying feature", func() {
+				Expect(createNs.Apply()).To(Succeed())
+				namespace, err := getNamespace("embedded-test-ns")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(namespace).ToNot(BeNil())
+			})
```
Comment (bartoszmajsak) [inline tests/integration/features/features_int_test.go:217]: Same as before. We are missing `then` block.

File: tests/integration/features/features_int_test.go [hunk 7]
```diff
@@ -127,6 +135,119 @@ var _ = Describe("preconditions", func() {
...
+		})
+
+		It("should be able to source manifests from an actual file system", func() {
```
Comment (bartoszmajsak) [inline tests/integration/features/features_int_test.go:220]: (nit)

```suggestion
		It("should source manifests from a specified temporary directory within the file system", func() {
```

File: tests/integration/features/features_int_test.go [hunk 8]
```diff
@@ -127,6 +135,119 @@ var _ = Describe("preconditions", func() {
...
+
+			Expect(err).ToNot(HaveOccurred())
+
+			// when
+			By("verifying namespace is created after applying feature", func() {
+				Expect(createNs.Apply()).To(Succeed())
+				namespace, err := getNamespace("real-file-test-ns")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(namespace).ToNot(BeNil())
+			})
```
Comment (bartoszmajsak) [inline tests/integration/features/features_int_test.go:246]: again - missing `then` part :)

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]
Thanks @cam-garrison! I shared few ideas for improvement, but these are small things.

File: pkg/feature/manifest.go (hunk 2 above)
Comment (cam-garrison) [reply pkg/feature/manifest.go]: agree, done here [54f4f38](https://github.com/opendatahub-io/opendatahub-operator/pull/797/commits/54f4f3835be46134b21a2ec5dba81e2517bfe93d)

File: pkg/feature/manifest.go (hunk 1 above)
Comment (cam-garrison) [reply pkg/feature/manifest.go]: good catch, done here [54f4f38](https://github.com/opendatahub-io/opendatahub-operator/pull/797/commits/54f4f3835be46134b21a2ec5dba81e2517bfe93d)

File: tests/integration/features/features_int_test.go (hunk 8 above)
Comment (cam-garrison) [reply tests/integration/features/features_int_test.go]: fixed! 
740741b

File: tests/integration/features/features_int_test.go (hunk 6 above)
Comment (cam-garrison) [reply tests/integration/features/features_int_test.go]: 740741b

File: tests/integration/features/features_int_test.go (hunk 4 above)
Comment (cam-garrison) [reply tests/integration/features/features_int_test.go]: 740741b added the name check here!

**Comment by bartoszmajsak** [comment]
Cluster seems to be borked:

> `level=error msg=Cluster initialization failed because one or more operators are not functioning properly.`

but let's try again.

/retest

File: pkg/feature/feature.go [hunk 9]
```diff
@@ -71,35 +74,32 @@ func (f *Feature) Apply() error {
...
 	// Process and apply manifests
-	for _, m := range f.manifests {
-		if err := m.processTemplate(f.Spec); err != nil {
+	for i, m := range f.manifests {
+		if err := m.process(f.fsys, f.Spec); err != nil {
 			return errors.WithStack(err)
 		}
 
-		log.Info("converted template to manifest", "feature", f.Name, "path", m.targetPath())
+		f.manifests[i] = m
```
Comment (zdtsw) [inline pkg/feature/feature.go:90]: why is this needed?

Comment (cam-garrison) [reply pkg/feature/feature.go:90]: since range() creates a copy of each element in the slice and we modify the copied manifest, we update the original slice with the modified manifest.

File: pkg/feature/raw_resources.go [hunk 10]
```diff
@@ -105,25 +94,16 @@ func (f *Feature) patchResourceFromFile(filename string) error {
 			Resource: strings.ToLower(u.GroupVersionKind().Kind) + "s",
 		}
 
-		// Convert the patch from YAML to JSON
-		patchAsJSON, err := yaml.YAMLToJSON(data)
+		// Convert the individual resource patch from YAML to JSON
```
[the remark this answers is not in the document]
Comment (cam-garrison) [reply pkg/feature/raw_resources.go:97]: I believe it's necessary as we want to convert to JSON in case of JSON merge patch or JSON patch formats used for patching - @bartoszmajsak can you confirm or am I off base here?

File: pkg/feature/feature.go (hunk 9 above)
Comment (israel-hdez) [reply pkg/feature/feature.go:90]: Can we do `f.manifests[i].process` within the `if` rather than processing in the copy?

File: pkg/feature/manifest.go [hunk 11]
```diff
@@ -1,76 +1,101 @@
...
+		return nil, err
 	}
 
 	return manifests, nil
 }
 
+func createManifestFrom(path string) manifest {
+	basePath := filepath.Base(path)
+	m := manifest{
+		name:     basePath,
```
Comment (israel-hdez) [inline pkg/feature/manifest.go:53]: Wouldn't it make sense to add a reference to the file system where the manifest is located? That way, `process` won't need the fsys argument, and it will be cleaner if we ever use several `fs.FS`.

Comment (israel-hdez) [reply pkg/feature/manifest.go:53]: Although... well... the FS seems to be stored in the `Feature`.

File: pkg/feature/raw_resources.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (zdtsw) [reply pkg/feature/raw_resources.go:97]: you are right
by using `MergePatchType` it uses "application/merge-patch+json" which need to be json format.

File: pkg/feature/manifest.go (hunk 11 above)
Comment (bartoszmajsak) [reply pkg/feature/manifest.go:54]: In fact, this is an aspect of manifest (the source of it), not a feature per se, so it should belong to the manifest struct instead. I took the liberty and reworked it in 588e61c /cc @cam-garrison

Comment (bartoszmajsak) [reply pkg/feature/manifest.go:54]: Thanks for the great suggestion @israel-hdez

File: pkg/feature/feature.go (hunk 9 above)
Comment (bartoszmajsak) [reply pkg/feature/feature.go:88]: I think we can simply do 

```golang
for i := range f.manifests {
	if err := f.manifests[i].process(f.Spec); err != nil {
		return errors.WithStack(err)
	}
}
```

Comment (cam-garrison) [reply pkg/feature/feature.go]: agreed, thx! 
3c6faeb

**Review by zdtsw — APPROVED** [review]

**Comment by cam-garrison** [comment]
``` 
* could not run steps: step [release:latest] failed: release "release-latest" failed: pod pending for more than 30m0s: containers have not started in 39m47.575382799s: ci-scheduling-dns-wait, place-entrypoint, release, sidecar: 
* Container release is not ready with reason PodInitializing
```

/retest

**Review by bartoszmajsak — APPROVED** [review]
